### PR TITLE
Topic/annotated abts

### PR DESCRIPTION
**Pull request by pchiusano**
The `ABT.Term f` is now `Term f a`, where `a` is the type of annotations. Change was very surgical, and with the pattern synonyms, almost no code outside of `ABT.hs` needed to change.

There's also:

* a `Functor (ABT.Term f)` so you can use `fmap` to alter the annotations throughout the tree
* `annotation :: ABT.Term f a -> a` to get the annotation
* `annotate :: a -> ABT.Term f a -> ABT.Term f a` to set the annotation
* A number of functions in `ABT.hs` have been generalized to work with annotated terms. There might be a few that could be usefully generalized that I haven't done yet.

The `Unison.Term.Term` type alias is now an alias for `ABT.Term F ()`. I did it this way just to make the upgrade super easy, and also because the vast majority of the code doesn't deal with these annotations at all (and wouldn't have enough information to propagate the info in the annotations). There's `AnnotatedTerm` if you want to allow non-`()` annotations, likewise there is `Type` and `AnnotatedType`.

I'm thinking of merging this as is, and if we encounter places where it makes sense to generalize from `Type` to `AnnotatedType` or `Term` to `AnnotatedTerm`, then we can do that as needed. However, since it's often more complicated to generalize, I'd like to only do this in places where there's a compelling reason to do so, not just because.

I considered using a typeclass for handling propagation of annotations up the tree, but decided that was a bad idea. Annotations must be set explicitly, and there's no meaning attached to the annotations, you can stick anything in there.

Review by @int-index and @Ericson2314 

I'll probably merge sometime tomorrow if I don't think of any problems with this or neither of you has serious objections.

### CODE REVIEW

**Comment by int-index** [comment]
No objections.

**Comment by Ericson2314** [comment]
Mmm, I'd prefer some sort of type class for the fixpoints. The traversal functions can be generic on it, while code generating ASTs can just the concrete old `Term`. I don't think `Cofree` is enough to handle all possible uses (consider `Free` for commenting an AST node).

**Comment by pchiusano** [comment]
@Ericson2314 Not totally sure what you mean, can you elaborate?

**Comment by Ericson2314** [comment]
I'll try. :)

We both independently discovered this http://brianmckenna.org/blog/type_annotation_cofree , if I remember. To start with, I think it is good to follow that separate the fix-point type from other stuff, so so even  `Term` is not recursive. Second, I think it is good to have `Mu` and `Cofree`, and not `Cofree ()`, because in general we won't be able to force all IRs we want to use a single fix-point type. My comments example was was wrong as `Cofree (Maybe String/Text/String) AstFunctor` works file. An actually example is for incomplete expressions using `Free Hole AstFunctor`

**Comment by pchiusano** [comment]
@Ericson2314 do you mean like this:

```Haskell
data ABT f a
  = Var V
  | Cycle a
  | Abs V a
  | Tm (f a) deriving (Functor, Foldable, Traversable)

data Term' f r = Term' { freeVars :: Set V, out :: ABT f r }

-- An ABT tree annotated with `a` values at every level
type Term f a = Cofree (Term' f) a

-- An ABT tree with a hole of type `a`
type Term1 f a = Free (Term' f) a
```

**Comment by Ericson2314** [comment]
Yeah, exactly.

The class part of my original comment was to try to add some way to reverse that was agnostic to `Mu`/`Cofree`/`Free`, but we don't need to worry about that for now.

**Comment by pchiusano** [comment]
@Ericson2314 cool. I'll probably go ahead and do that, though `Term` needs to be newtyped in order to give all the typeclass instances. I don't think I can get to this until next week though, I have other stuff I am working on today and will be busy this weekend.

**Comment by Ericson2314** [comment]
I guess I can take a look

**Comment by pchiusano** [comment]
@Ericson2314 Actually are you opposed to merging as is? I think that will unblock @int-index and switching to what I gave above should be just as surgical and can be done later. I'd like to make sure this change doesn't mess with anything else.

**Comment by Ericson2314** [comment]
Good point. Go for it!

**Comment by Ericson2314** [comment]
Only caveat is that many things there are currently `Cofree () ...` should probably be written with `Mu`/, in which case they might sort of change back.

**Comment by pchiusano** [comment]
Okay cool, merging. That makes sense about using `Mu` for the existing stuff rather than `Cofree ()`. The reason I think it won't really matter is that all the code uses the patterns defined in `ABT`, `Term`, and `Type`, so that level of indirection is what makes it not really require changes outside those pattern definitions.

I'll also preserve the `annotation` and `annotate` functions, so if you are using those you should be good regardless of the underlying representation. @int-index, I would use these (and the other accessor functions) rather than direct raw pattern matching on the representation. If you use raw pattern matching you might have to update some code later.

**Comment by pchiusano** [comment]
@Ericson2314 played with this a bit. The issue with:

```Haskell
data Term' f r = Term' { freeVars :: Set V, out :: ABT f r }

newtype Fix f = Fix (f (Fix f))

newtype AnnotatedTerm f a = AnnotatedTerm (Cofree (Term' f) a)
newtype Term f = Term (Fix (Term' f))
```

is that now, anyplace we want to work with terms, we either need to pick one of `Fix` or `Cofree`, or come up with some typeclass to abstract over these fixpoints if we don't care which it is. Okay, so we can introduce something like: 

```Haskell
class Fixpoint t f | t -> f where
  unroll :: t -> f t
  reroll :: t -> (f t -> f t) -> t
```

But that's not sufficient, since we also have stuff like `ToJSON`, `Eq`, and `Serial` instances for ABTs. So we'll need to come up with some more ad hoc operations to support those. If we don't introduce a class and just pick one of `Fix` or `Cofree` in all the functions, well then we're seem to be worse off than the current, which can at least be polymorphic in the annotations if it doesn't care about them.

Basically, I see this adding complexity that doesn't really pay for itself---I don't have any really compelling use cases to justify it right now. I am happy to revisit this later if we do come up with something, but for now I'd like to leave it as is.

File: node/src/Unison/ABT/Extra.hs [hunk 1]
```diff
@@ -21,10 +21,12 @@ import qualified Data.Map as Map
 import qualified Data.Vector as Vector
 import qualified Unison.Digest as Digest
 
-hash :: forall f . (Foldable f, Digest.Digestable1 f) => Term f -> Digest.Hash
+-- | We ignore annotations in the `Term`, as these should never affect the
+-- meaning of the term.
```
Comment (Ericson2314) [inline node/src/Unison/ABT/Extra.hs]: We rely more heavily on the type-checking algorithm to give meaning to our programs if we do this. Part of me would rather have the fully-elaborated AST be the "final truth" of the program that we persist. Text-based software development problem *doesn't* do this today because it would make refactoring that much harder, but the ideas you propose in your last post may alleviate this. Furthermore we can distinguish inferred and explicit types if we need.

File: shared/src/Unison/ABT.hs [hunk 2]
```diff
@@ -239,25 +272,33 @@ instance (Foldable f, Functor f, Eq1 f) => Eq (Term f) where
     go (Tm f1) (Tm f2) = f1 ==# f2
     go _ _ = False
 
-instance J.ToJSON1 f => ToJSON (Term f) where
```
Comment (Ericson2314) [inline shared/src/Unison/ABT.hs]: Now that we are using ghcjs, might we just auto derive these? It's one less parsing/serializing/marshaling nuisance to avoid -- might as well dogfood our own philosophy!

**Comment by Ericson2314** [comment]
@pchiusano First of all, sorry for responding a few days later.

I had made some progress by have separate `Roll` and `Unroll` classes. I'll admit however the result still felt awkward and over-abstracted-- -basically code that confirmed all the negative Haskell stereotypes. So yeah, let's ditch the classes for now. 

> If we don't introduce a class and just pick one of Fix or Cofree in all the functions, well then we're seem to be worse off than the current, which can at least be polymorphic in the annotations if it doesn't care about them.

So while the operations may not care about the annotations, the annotations may care about the operations---i.e. even leaving them in place somehow breaks some invariant the annotations were supposed to represent. Avoiding annotation-polymorphic functions, and *especially* instances, allows the annotation definer to redefine the operations preserving the invariants without having to resort to yet another newtype.

File: node/src/Unison/ABT/Extra.hs (hunk 1 above)
Comment (pchiusano) [inline node/src/Unison/ABT/Extra.hs]: I was not positive about this decision, but the idea of leaving the annotation out of the hashes is that the annotation does not affect the meaning of a term, and the hash is meant to uniquely identify a meaning. Same reason that the free variables set doesn't affect hashing - it's just bookkeeping info.

Terms are already tied to a typechecker anyway, the typechecker is telling us what terms even make sense to run.

Anyway, I am still kind of skeptical that storing inferred types everywhere is a good idea. Besides the inefficiency, it also gets out of date in really unpredictable ways and seems error prone to use. You might modify a subterm, and now subterms very far from your original edit have incorrect annotations, depending on how typing info flows. Having to worry about whether I'm doing that does not sound fun...

If you only recreate the info on demand based on the current state of the world, there's no way it can ever be out of sync! :) As long as you can service these queries in an amount of time that a human perceives as being instantaneous, you are good.

Anyway, I don't want to discourage you from trying out building a fully-annotated tree if you're super interested in that. I'm just not personally sold on it... :) I see it being potentially useful as more of an efficiency thing - you ask the node to give you back an annotated tree, which lets you hover over multiple subexpressions to view their types, without a roundtrip to the node each time.

File: shared/src/Unison/ABT.hs (hunk 2 above)
Comment (pchiusano) [inline shared/src/Unison/ABT.hs]: Well, the thing that is a little weird here is that we are stripping out the free variables set from the JSON (and binary) formats, and rederiving that info on the other end...

Comment (Ericson2314) [inline shared/src/Unison/ABT.hs]: How about converting to a type without the free variable sets and then deriving to/from JSON on that?
